Framework/Source/Graphics/Program/ProgramReflection.cpp

```diff
@@ -0,0 +1,1190 @@
+                }
+                break;
+            case 2:
+                switch (columns)
+                {
+                case 2:
+                    return ReflectionBasicType::Type::Float2x2;
+                case 3:
+                    return ReflectionBasicType::Type::Float2x3;
+                case 4:
+                    return ReflectionBasicType::Type::Float2x4;
+                }
+                break;
+            case 3:
+                switch (columns)
+                {
+                case 2:
+                    return ReflectionBasicType::Type::Float3x2;
+                case 3:
+                    return ReflectionBasicType::Type::Float3x3;
+                case 4:
+                    return ReflectionBasicType::Type::Float3x4;
+                }
+                break;
+            case 4:
+                switch (columns)
+                {
+                case 2:
+                    return ReflectionBasicType::Type::Float4x2;
+                case 3:
+                    return ReflectionBasicType::Type::Float4x3;
+                case 4:
+                    return ReflectionBasicType::Type::Float4x4;
+                }
+                break;
+            }
+        }
+
+        should_not_get_here();
+        return ReflectionBasicType::Type(-1);
+    }
+
+    static ReflectionResourceType::StructuredType getStructuredBufferType(TypeReflection* pSlangType)
+    {
+        auto invalid = ReflectionResourceType::StructuredType::Invalid;
+
+        if (pSlangType->getKind() != TypeReflection::Kind::Resource)
+            return invalid; // not a structured buffer
+
+        if (pSlangType->getResourceShape() != SLANG_STRUCTURED_BUFFER)
+            return invalid; // not a structured buffer
+
+        switch (pSlangType->getResourceAccess())
+        {
+        default:
+            should_not_get_here();
+            return invalid;
+
+        case SLANG_RESOURCE_ACCESS_READ:
+            return ReflectionResourceType::StructuredType::Default;
+
+        case SLANG_RESOURCE_ACCESS_READ_WRITE:
+        case SLANG_RESOURCE_ACCESS_RASTER_ORDERED:
+            return ReflectionResourceType::StructuredType::Counter;
+        case SLANG_RESOURCE_ACCESS_APPEND:
+            return ReflectionResourceType::StructuredType::Append;
+        case SLANG_RESOURCE_ACCESS_CONSUME:
+            return ReflectionResourceType::StructuredType::Consume;
+        }
+    };
+
+    ReflectionVar::SharedPtr reflectVariable(VariableLayoutReflection* pSlangLayout, const ReflectionPath* pPath);
+    ReflectionType::SharedPtr reflectType(TypeLayoutReflection* pSlangType, const ReflectionPath* pPath);
+
+    // Once we've found the path from the root down to a particular leaf
+    // variable, `getDescOffset` can be used to find the final summed-up descriptor offset of the element
+    static uint32_t getDescOffsetFromPath(const ReflectionPath* pPath, uint32_t arraySize, SlangParameterCategory category)
+    {
```

Comment:
If I understand correctly, this is to handle cases where you have:

```glsl
texture2d t[20];
```

and you need to find which descriptor correspondings to `t[3]`, but Slang is only telling you that it gave the whole *array* `binding = 0` or whatever.

It probably makes sense for Slang to track two different things for Vulkan shader parameters: both the "bindings"
 (where an array only consumes one binding, no matter how many elements) and the "descriptors" (where an array consumes one or more descriptors per element). That would in principle let you use the same logic as, e.g., `getRegisterIndexFromPath`, right?